src/Pos/DB/Block/Block.hs

```diff
@@ -0,0 +1,302 @@
+import qualified Data.ByteString.Lazy      as BSL
+import           Data.Default              (Default (def))
+import           Formatting                (build, formatToString, int, sformat, (%))
+import           System.FilePath           ((</>))
+import           System.IO                 (IOMode (ReadMode), SeekMode (AbsoluteSeek),
+                                            hFileSize, hSeek, withFile)
+import           Universum
+
+import           Pos.Binary.Block          ()
+import           Pos.Binary.Class          (Bi, FixedSizeInt (..), WithLength (..),
+                                            decodeFull, encodeStrict)
+import           Pos.Binary.DB             ()
+import           Pos.Block.Types           (Blund, Undo (..))
+import           Pos.Constants             (maxBlundFileSize)
+import           Pos.Crypto                (shortHashF)
+import           Pos.DB.Block.Aux          (BlundAux (..))
+import           Pos.DB.Class              (MonadDB, getBlockIndexDB, getNodeDBs)
+import           Pos.DB.Error              (DBError (DBMalformed))
+import           Pos.DB.Functions          (rocksGetBi, rocksPutBi)
+import           Pos.DB.Types              (blockDataDir)
+import           Pos.Ssc.Class.Helpers     (SscHelpersClass)
+import           Pos.Types                 (Block, BlockHeader, GenesisBlock,
+                                            HasDifficulty (difficultyL), HasPrevBlock,
+                                            HeaderHash, genesisHash, headerHash,
+                                            prevBlockL)
+import qualified Pos.Types                 as T
+import           Pos.Util                  (maybeThrow)
+import           Pos.Util.Chrono           (NewestFirst (..))
+
+-- | Get block with given hash from Block DB.
+getBlock
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (Block ssc))
+getBlock hh = getBi (blockAuxKey hh) >>= \case
+    Nothing -> pure Nothing
+    Just BlundAux{..} -> blundDataPath blundFileNum >>= fmap Just . getData blockOffset blockLen
+
+-- | Returns header of block that was requested from Block DB.
+getBlockHeader
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (BlockHeader ssc))
+getBlockHeader = getBi . blockIndexKey
+
+-- | Get undo data for block with given hash from Block DB.
+getUndo :: (MonadDB m) => HeaderHash -> m (Maybe Undo)
+getUndo hh = getBi (blockAuxKey hh) >>= \case
+    Nothing -> pure Nothing
+    Just BlundAux{..} -> blundDataPath blundFileNum >>= fmap Just . getData undoOffset undoLen
+
+-- | Retrieves block and undo together.
+getBlockWithUndo
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (Block ssc, Undo))
+getBlockWithUndo x =
+    runMaybeT $ (,) <$> MaybeT (getBlock x) <*> MaybeT (getUndo x)
+
+-- | Put given block, its metadata and Undo data into Block DB.
+putBlund
+    :: (SscHelpersClass ssc, MonadDB m)
+    => Block ssc -> Undo -> m ()
```

Comment:
Do I understand correctly that if program stops somewhere inside this function (or exception is thrown), there may be some redundant data in files, but nothing will be broken?

If that's true, I think it's acceptable, but can we make it better? I. e. ensure that we store each block only once. I assume there will be fix which will make this function idempotent as long as it terminates normally (e. g. one can check that value exists in block index). But if program stops after (or inside) `putData`, we will put redundant data into block files.